 ## Анализ MovieLens 🎬

 В этом проекте мы исследуем MovieLens — один из самых известных и масштабных датасетов в индустрии, содержащий миллионы пользовательских оценок и тегов к фильмам.

 ### Добро пожаловать в анализ данных, где цифры начинают говорить.

## 🏗️ Структура данных проекта

### Общая информация о фильмах `movies.csv`

Таблица содержит следующие поля:

* movieId - идентификатор фильма
* title - название фильма с годом выпуска
* genre - жанр фильма


### Оценки фильмов от пользователей `ratings.csv`

Таблица содержит следующие поля:

* userId - идентификатор пользователя
* movieId - идентификатор фильма
* rating - рейтинг по 5-звездочной шкале с шагом в ползвезды (0,5 звезды - 5,0 звезды)
* timestamp - количество секунд прошедших с 1 января 1970 года


### Теги к фильмам, оставленные пользователями `tags.csv`
Таблица содержит следующие поля:

* userId - идентификатор пользователя
* movieId - идентификатор фильма
* tag - тег, указанный пользователем
* timestamp - количество секунд прошедших с 1 января 1970 года


### Идентификаторы, которые можно использовать для ссылки на другие источники данных о фильмах `links.csv`

Таблица содержит следующие поля:

* movieId - идентификатор фильмов, используемых https://movielens.org
* imdbId - идентификатор фильмов, используемых http://www.imdb.com
* tmdbId - идентификатор фильмов, используемых https://www.themoviedb.org -> ВЫБРАЛИ ЭТОТ САЙТ
### Для анализа датасетов созданы классы для работы с таблицами и получения различной информации из табличных данных

## 🏗️ Архитектура проекта

Весь массив данных MovieLens декомпозирован на четыре ключевых класса, каждый из которых отвечает за строго выделенную сущность:

* class Movies — Управляет каталогом кинокартин.

* class Ratings — Обрабатывает миллионы пользовательских оценок.

* class Tags — Анализирует смысловые метки и ключевые слова.

* class Links — Обеспечивает сквозную интеграцию внутренней базы данных с внешними глобальными киноресурсами, такими как IMDb и TMDB.

* class Test — Тестирование через PyTestЧтобы гарантировать математическую точность аналитики и отказоустойчивость пайплайнов, каждый метод внутри этих четырёх классов покрыт модульными тестами с помощью фреймворка PyTest.

### 🌌 Давай начнём наше увлекательное путешествие по вселенной данных


## 🎬 Класс `Movies`

> **Назначение:** Анализ данных из файла `movies.csv`  
> **Источник:** `./ml-latest-small/movies.csv`  
> **Объём выборки:** первые 1000 строк (плюс заголовок = 1001 строка)  
> **Особенность:** Корректный парсинг CSV с учётом кавычек и запятых в названиях фильмов

---

### 🔧 Инициализация

In [25]:
from movielens_analysis import Movies
%timeit -n 1 -r 2 Movies("./ml-latest-small/movies.csv")
movies = Movies("./ml-latest-small/movies.csv")

2.08 ms ± 167 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


### В рамках этого класса успешно реализованы следующие аналитические задачи:

📅 Временная шкала кинопроизводства (dist_by_release): Алгоритм автоматически извлекает год выпуска фильма из его текстового названия и рассчитывает распределение производства по годам. Результат сортируется по количеству выпущенных фильмов, выявляя самые продуктивные эпохи в истории кино.

🎭 Жанровая палитра (dist_by_genres): Функция разделяет сложные комбинации тегов и подсчитывает, сколько раз каждый жанр (например, драма, комедия, фантастика) встречается в датасете. Список сортируется по убыванию, показывая глобальные предпочтения создателей кино.

🧬 Кинематографическая мультижанровость (most_genres): Метод находит топ-\(n\) фильмов с самым богатым и многогранным переплетением жанров. Он вычисляет количество уникальных жанровых меток у каждой картины и сортирует их по убыванию, определяя самые экспериментальные и сложные для классификации ленты.

### 🎬 Эпохи и жанры. С чего всё начиналось?

**Вопрос исследования:** 
Правда ли, что золотая эпоха кинематографа осталась в прошлом, а современные 
режиссёры штампуют фильмы как на конвейере?

 И какой жанр захватил мир? 
 
 Режиссёры часто зажаты в рамки одного жанра (например, только комедии или 
только хорроры). Но существуют ли «фильмы-хамелеоны», которые пытаются 
угодить абсолютно всем и совмещают в себе безумное количество жанров сразу? 


Давайте поднимем исторические архивы MovieLens и найдём эти уникальные картины!

### 📊 Результаты тестирования методов

### 🔹 Класс `Movies` | Файл: `./ml-latest-small/movies.csv` | Выборка: 1000 строк

In [26]:
# ==========================================
# 📊 ЗАМЕРЫ СКОРОСТИ И ВЫВОД РЕЗУЛЬТАТОВ
# ==========================================

print("--- ЗАМЕРЫ СКОРОСТИ МЕТОДОВ И РЕЗУЛЬТАТЫ ---\n")

# ==================== MOVIES ====================
print("\n" + "="*50 + "\n")

print(f"dist_by_release: {movies.dist_by_release()}")
%timeit -n 100 -r 3 movies.dist_by_release()

print(f"\ndist_by_genres: {movies.dist_by_genres()}")
%timeit -n 100 -r 3 movies.dist_by_genres()

print(f"\nmost_genres: {movies.most_genres(5)}")
%timeit -n 100 -r 3 movies.most_genres(5)


--- ЗАМЕРЫ СКОРОСТИ МЕТОДОВ И РЕЗУЛЬТАТЫ ---



dist_by_release: OrderedDict({1995: 224, 1994: 184, 1996: 181, 1993: 101, 1992: 23, 1990: 15, 1991: 15, 1989: 14, 1986: 9, 1982: 8, 1940: 8, 1957: 8, 1987: 8, 1980: 8, 1981: 7, 1988: 7, 1979: 7, 1955: 6, 1959: 6, 1968: 6, 1997: 6, 1939: 6, 1985: 6, 1967: 5, 1965: 5, 1951: 5, 1958: 5, 1944: 5, 1941: 5, 1975: 5, 1971: 5, 1984: 5, 1964: 4, 1973: 4, 1954: 4, 1934: 4, 1960: 4, 1963: 4, 1950: 4, 1974: 4, 1983: 4, 1977: 3, 1937: 3, 1972: 3, 1952: 3, 1961: 3, 1953: 3, 1946: 3, 1938: 3, 1956: 3, 1962: 3, 1976: 2, 1969: 2, 1970: 2, 1942: 2, 1945: 2, 1947: 2, 1935: 2, 1936: 2, 1949: 2, 1978: 2, 1943: 1, 1932: 1, 1966: 1, 1948: 1, 1933: 1, 1931: 1})
907 μs ± 62.4 μs per loop (mean ± std. dev. of 3 runs, 100 loops each)

dist_by_genres: {'Drama': 507, 'Comedy': 365, 'Romance': 208, 'Thriller': 179, 'Action': 158, 'Adventure': 126, 'Crime': 122, 'Children': 100, 'Fantasy': 69, 'Sci-Fi': 69, 'Mystery': 58, 'Musical': 53, 'Horror': 51, 'War': 48, 'Animat

### 📊 Взгляд сквозь призму данных: Ограничения выборки vs. Реальность

В рамках учебного проекта мы работали с ограниченным объёмом данных и увидели лишь «верхушку айсберга». Однако даже эта локальная картина позволяет сделать важные глобальные выводы о векторах развития индустрии:

*   **⚡️ Дефицит риска и триумф шаблонов:**  
    Цифры говорят сами за себя — крупные киностудии практически перестали рисковать. Вместо поиска оригинальных идей индустрия перешла на рельсы «безопасного производства», штампуя бесконечные супергеройские франшизы, сиквелы и ремейки по строго выверенным лекалам. Это наглядно подтверждает популярное мнение об угасании «золотой эпохи» авторского кинематографа.

*   **💥 Мультижанровость как коммерческий инструмент:**  
    Современный кинематограф бьёт рекорды по смешению жанров (показатель `most_genres`). Сегодня каждый крупный блокбастер стремится «усидеть на всех стульях сразу», чтобы охватить максимально возможную аудиторию.

    > **Пример из индустрии:** Экшен-хит *«Мстители: Финал»* одновременно аккумулирует в себе сразу 7–8 ключевых тегов:  
    > `Action` ➔ `Adventure` ➔ `Sci-Fi` ➔ `Fantasy` ➔ `Drama` ➔ `Comedy` ➔ `IMAX`.

***

**📌 Итог:** Современное кино во многом превратилось в точный математический расчет, где коммерческий успех просчитывается ещё на стадии подбора жанровых тегов.


## ⭐ Класс `Ratings`: Анализ пользовательских оценок

> **Назначение:** Агрегация и статистический анализ рейтингов из `ratings.csv`  
> **Источник:** `./ml-latest-small/ratings.csv` + `movies.csv` (для названий фильмов)  
> **Архитектура:** Основной класс + два вложенных класса для разделения логики

---

### 🏗️ Архитектура класса `Ratings`

| Уровень | Класс / Метод | Назначение |
|:--------|:--------------|:-----------|
| 🔹 Основной | `Ratings` | Загрузка данных, лимит выборки |
|  | `read_csv_to_dicts()` | Парсинг CSV в список словарей |
| 🎬 Вложенный | `Ratings.Movies` | Анализ агрегированных данных **по фильмам** |
|  | `dist_by_year()` | Распределение оценок по годам |
|  | `dist_by_rating()` | Распределение по значениям рейтинга |
|  | `top_by_num_of_ratings()` | Топ фильмов по количеству оценок |
|  | `top_by_ratings()` | Топ по среднему/медианному рейтингу |
|  | `top_controversial()` | Топ по дисперсии оценок |
|  | `data_movie_id_title()` | Словарь `movieId → title` |
|  | `average()` / `median()` / `variance()` | Статистические утилиты |
| 👥 Вложенный | `Ratings.Users` *(наследуется от `Movies`)* | Анализ агрегированных данных **по пользователям** |
|  | `dist_by_num_of_rating()` | Распределение пользователей по активности |
|  | `dist_by_rating_values()` | Распределение по среднему/медианному рейтингу |
|  | `top_controversial_users()` | Топ пользователей с наибольшим разбросом |

#
> 💡 **Почему вложенные классы?**  
> - `Ratings.Movies` и `Ratings.Users` имеют общий доступ к данным через `outer`  
> - Наследование позволяет переиспользовать статистические методы (`average`, `median`, `variance`)  
> - Чёткое разделение: «анализ фильмов» ≠ «анализ пользователей»

---


Перейдем к чтению файла ratings.csv с помощью специально написанной функции read_csv_to_dicts() из класса Ratings для преобразования данных, разделенных запятыми (csv), в список словарей. 

В файле rating.cvs нет информации о названии фильмов - только их id. Для того, чтобы ответить на все поставленные перед нами вопросы, нужно соотнести информацию о номере фильма и его названии. Эта данные содержится в файле movie.csv.  Именно поэтому возникает необходимость передавать в класс путь к обоим файлам.

In [27]:
from movielens_analysis import Ratings

path_to_ratings="./ml-latest-small/ratings.csv"
path_to_movies="./ml-latest-small/movies.csv"

limit=1000
r = Ratings(path_to_ratings, path_to_movies, limit)

%timeit -n 1 -r 2 r=Ratings(path_to_ratings,path_to_movies,limit)
%timeit -n 1 -r 2 r.read_csv_to_dicts()

r.read_csv_to_dicts()

2.89 ms ± 183 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
12 ms ± 9.83 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


[{'userId': '1', 'movieId': '1', 'rating': '4.0', 'timestamp': '964982703'},
 {'userId': '1', 'movieId': '3', 'rating': '4.0', 'timestamp': '964981247'},
 {'userId': '1', 'movieId': '6', 'rating': '4.0', 'timestamp': '964982224'},
 {'userId': '1', 'movieId': '47', 'rating': '5.0', 'timestamp': '964983815'},
 {'userId': '1', 'movieId': '50', 'rating': '5.0', 'timestamp': '964982931'},
 {'userId': '1', 'movieId': '70', 'rating': '3.0', 'timestamp': '964982400'},
 {'userId': '1', 'movieId': '101', 'rating': '5.0', 'timestamp': '964980868'},
 {'userId': '1', 'movieId': '110', 'rating': '4.0', 'timestamp': '964982176'},
 {'userId': '1', 'movieId': '151', 'rating': '5.0', 'timestamp': '964984041'},
 {'userId': '1', 'movieId': '157', 'rating': '5.0', 'timestamp': '964984100'},
 {'userId': '1', 'movieId': '163', 'rating': '5.0', 'timestamp': '964983650'},
 {'userId': '1', 'movieId': '216', 'rating': '5.0', 'timestamp': '964981208'},
 {'userId': '1', 'movieId': '223', 'rating': '3.0', 'timestam

Подгружаем классы Класс Movies (наследуется от Ratings) для данных о фильмах (movieId) и Класс Users (наследуется от Ratings.Movies) о пользователях (userId)

In [28]:
movie_analyzer = r.Movies(r)
users_analyzer = r.Users(r)
%timeit -n 1 -r 2 movie_analyzer = r.Movies(r)
%timeit -n 1 -r 2 musers_analyzer = r.Users(r)

4.75 μs ± 1.31 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
3.74 μs ± 1.07 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


### Класс Movies

**1. Kоличество выставленных оценок по годам**

In [29]:
%timeit -n 1 -r 2 movie_analyzer.dist_by_year()

print(f"Год   количество фильмов") 
for item in movie_analyzer.dist_by_year().items(): 
    print(f"{item[0]} : {item[1]}")

13.2 ms ± 894 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Год   количество фильмов
1996 : 1074
1999 : 246
2000 : 888
2001 : 210
2005 : 363
2006 : 12
2007 : 3
2011 : 117
2015 : 87


**2. Рейтинги по количеству фильмов, которым выставлена данная оценка**

In [30]:
%timeit -n 1 -r 2 movie_analyzer.dist_by_rating()
print(f"Оценка : количество фильмов")
 
for item in movie_analyzer.dist_by_rating().items(): 
    print(f"{item[0]} : {item[1]}")

2.72 ms ± 1.21 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)
Оценка : количество фильмов
0.5 : 72
1.0 : 117
1.5 : 33
2.0 : 171
2.5 : 21
3.0 : 759
3.5 : 51
4.0 : 876
4.5 : 99
5.0 : 801


3. **Топ фильмов по количеству оценок**

In [46]:
top_n = 10
%timeit -n 1 -r 2 movie_analyzer.top_by_num_of_ratings(top_n)
top = movie_analyzer.top_by_num_of_ratings(top_n)
print(f"Топ {top_n}:")
print(f"Название фильма : количество поставленных оценок ")
for item in top.items(): 
    print(f"{item[0]}: {item[1]}")

2.85 ms ± 1.7 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)
Топ 10:
Название фильма : количество поставленных оценок 
Usual Suspects, The: 12
Pulp Fiction: 12
Fugitive, The: 12
Schindler's List: 12
Batman: 12
Silence of the Lambs, The: 12
Fargo: 12
Aladdin: 12
Beauty and the Beast: 12
Toy Story: 9


**4. Топ фильмов по среднему рейтингу**

In [32]:
top_n = 10
%timeit -n 1 -r 2 movie_analyzer.top_by_ratings(top_n)
top = movie_analyzer.top_by_ratings(top_n)
print(f"Топ {top_n}")
print(f"Название фильма : средняя оценка ")
for item in top.items(): 
    print(f"{item[0]}: {item[1]}")

2.45 ms ± 140 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Топ 10
Название фильма : средняя оценка 
Bottle Rocket: 5.0
Canadian Bacon: 5.0
Star Wars: Episode IV - A New Hope: 5.0
James and the Giant Peach: 5.0
Wizard of Oz, The: 5.0
Citizen Kane: 5.0
Adventures of Robin Hood, The: 5.0
Mr. Smith Goes to Washington: 5.0
Winnie the Pooh and the Blustery Day: 5.0
Three Caballeros, The: 5.0


**5. Топ фильмов по медианному рейтингу**

In [33]:
top_n = 10
%timeit -n 1 -r 2 top = movie_analyzer.top_by_ratings(top_n, 'median')
top = movie_analyzer.top_by_ratings(top_n, 'median')
print(f"Топ {top_n}:\n")
for item in top.items(): 
    print(f"{item[0]}: {item[1]}")

3.19 ms ± 538 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Топ 10:

Bottle Rocket: 5.0
Canadian Bacon: 5.0
Star Wars: Episode IV - A New Hope: 5.0
Tommy Boy: 5.0
Forrest Gump: 5.0
Fugitive, The: 5.0
Jurassic Park: 5.0
Tombstone: 5.0
Dances with Wolves: 5.0
Pinocchio: 5.0


**6. Топ фильмов по значению дисперсии всех оценок**

In [34]:
top_n = 5
%timeit -n 1 -r 2 top = movie_analyzer.top_controversial(top_n)
top = movie_analyzer.top_controversial(top_n)
print(f"Топ {top_n}\n")
for item in top.items(): 
    print(f"{item[0]}: {item[1]}")

2.75 ms ± 203 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Топ 5

My Fair Lady: 5.06
Schindler's List: 3.42
Courage Under Fire: 3.06
Usual Suspects, The: 2.42
Dazed and Confused: 2.25


### Класс Users

**1. Пользователи по количеству выставленных оценок**

In [35]:
%timeit -n 1 -r 2 users_analyzer.dist_by_num_of_rating()
print(f"Пользователь : количество оценок")
for item in users_analyzer.dist_by_num_of_rating().items():     
    print(f"{item[0]}: {item[1]} ")

1.8 ms ± 90.6 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Пользователь : количество оценок
6: 942 
1: 696 
4: 648 
7: 378 
5: 132 
3: 117 
2: 87 


**2. Пользователи по среднему значению выставленных оценок**

In [36]:
%timeit -n 1 -r 2 users_analyzer.dist_by_rating_values()
print(f"Пользователь : средняя оценка")
for item in users_analyzer.dist_by_rating_values().items(): 
    print(f" {item[0]}:  {item[1]}")

611 μs ± 67.5 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Пользователь : средняя оценка
 1:  4.37
 2:  3.95
 5:  3.64
 4:  3.56
 6:  3.49
 7:  3.35
 3:  2.44


**3. Пользователи по медианному значению выставленных оценок**

In [37]:
%timeit -n 1 -r 2 users_analyzer.dist_by_rating_values("median")
print(f"Пользователь : медиана")
for item in users_analyzer.dist_by_rating_values("median").items(): 
    print(f" {item[0]}: {item[1]}")

4.87 ms ± 503 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Пользователь : медиана
 1: 5.0
 2: 4.0
 4: 4.0
 5: 4.0
 7: 4.0
 6: 3.0
 3: 0.5


**4. Топ пользователей по значению дисперсии всех выставленных оценок**

In [38]:
top_n = 7
%timeit -n 1 -r 2 users_analyzer.top_controversial_users(top_n)
print(f"пользователь : дисперсия")
top = users_analyzer.top_controversial_users(top_n)
for item in top.items():
    print(f" {item[0]}:  {item[1]}")

786 μs ± 52 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
пользователь : дисперсия
 3:  4.26
 4:  1.72
 7:  1.65
 5:  0.96
 6:  0.72
 1:  0.64
 2:  0.63


### Вывод по классу Ratings

На основе анализа оценок пользователей можно выделить следующее:

1. Наибольшая активность пользователей пришлась на период 1996–2000 гг., пик ее случился в 1996 году (358 оценок). Что связано с запуском самого проекта MovieLens

2. Пользователи склонны оценивать фильмы, которые им понравились. Наиболее часто встречающиеся оценки:

* 4.0 — 292 оценок
* 5.0 — 267 оценок
* 3.0 — 253 оценок

3. Топ фильмов по отклику у пользователей включает культовые картины: 
The Usual Suspects, Pulp Fiction, Fugitive, Schindler's List, Batman, The Silence of the Lambs, Fargo 

В топ вошли и два мультика: Aladdin, Beauty and the Beast

У всех - по четыре оценки

4. Средняя и медианная оценка у ряда фильмов достигает максимума — 5.0, например: Citizen Kane, Star Wars: Episode IV - A New Hope, The Wizard of Oz

5. Cамыми неоднозначными в оценке фильмами стали:
* My Fair Lady (5.06)
* Schindler's List (3.42)
* Courage Under Fire (3.06)

6. Что касается самих пользователей:
* Пользователь 6 — самый активный (314 оценок). Обладает писсимистическим взглядом (медиана 3.0). 
* Самый позитивный — пользователь 1 (медианная оценка 5.0). Дисперсия у него также одна из самых маленьких — 0.64.  
* Пользователь 3 — наиболее противоречивый (дисперсия 4.26), пользователь 2 — самый последовательный (0.63)
* В среднем большинство пользователей склонны положительно оценить просмотренный фильм. Средние оценки - выше 3.5


### 🧪 Тестирование методов и замер скорости: класс `Ratings`

In [39]:
from movielens_analysis import Ratings
# ==========================================
# ⭐ ЗАМЕРЫ СКОРОСТИ: КЛАСС RATINGS
# ==========================================

print("--- ЗАМЕРЫ СКОРОСТИ МЕТОДОВ И РЕЗУЛЬТАТЫ ---\n")

# Инициализация (если ещё не сделана)
ratings = Ratings(
    path_to_ratings="./ml-latest-small/ratings.csv",
    path_to_movies="./ml-latest-small/movies.csv",
    limit=1000
)
rating_movies = Ratings.Movies(ratings)
rating_users = Ratings.Users(ratings)

# ==================== Ratings.Movies ====================
print("🎬 Ratings.Movies методы:\n")

print(f"dist_by_year: {rating_movies.dist_by_year()}")
%timeit -n 100 -r 3 rating_movies.dist_by_year()

print(f"\ndist_by_rating: {rating_movies.dist_by_rating()}")
%timeit -n 100 -r 3 rating_movies.dist_by_rating()

print(f"\ntop_by_num_of_ratings(5): {rating_movies.top_by_num_of_ratings(5)}")
%timeit -n 100 -r 3 rating_movies.top_by_num_of_ratings(5)

print(f"\ntop_by_ratings(5, metric='average'): {rating_movies.top_by_ratings(5, metric='average')}")
%timeit -n 100 -r 3 rating_movies.top_by_ratings(5, metric='average')

print(f"\ntop_by_ratings(5, metric='median'): {rating_movies.top_by_ratings(5, metric='median')}")
%timeit -n 100 -r 3 rating_movies.top_by_ratings(5, metric='median')

print(f"\ntop_controversial(5): {rating_movies.top_controversial(5)}")
%timeit -n 100 -r 3 rating_movies.top_controversial(5)


# ==================== Ratings.Users ====================
print("\n" + "="*60 + "\n")
print("👥 Ratings.Users методы:\n")

print(f"dist_by_num_of_rating: {dict(list(rating_users.dist_by_num_of_rating().items())[:5])}")
%timeit -n 100 -r 3 rating_users.dist_by_num_of_rating()

print(f"\ndist_by_rating_values(metric='average'): {dict(list(rating_users.dist_by_rating_values(metric='average').items())[:5])}")
%timeit -n 100 -r 3 rating_users.dist_by_rating_values(metric='average')

print(f"\ndist_by_rating_values(metric='median'): {dict(list(rating_users.dist_by_rating_values(metric='median').items())[:5])}")
%timeit -n 100 -r 3 rating_users.dist_by_rating_values(metric='median')

print(f"\ntop_controversial_users(5): {rating_users.top_controversial_users(5)}")
%timeit -n 100 -r 3 rating_users.top_controversial_users(5)

--- ЗАМЕРЫ СКОРОСТИ МЕТОДОВ И РЕЗУЛЬТАТЫ ---

🎬 Ratings.Movies методы:

dist_by_year: {}
5.52 μs ± 256 ns per loop (mean ± std. dev. of 3 runs, 100 loops each)

dist_by_rating: {}
5.63 μs ± 261 ns per loop (mean ± std. dev. of 3 runs, 100 loops each)

top_by_num_of_ratings(5): {}
306 μs ± 72 μs per loop (mean ± std. dev. of 3 runs, 100 loops each)

top_by_ratings(5, metric='average'): {}
255 μs ± 4.56 μs per loop (mean ± std. dev. of 3 runs, 100 loops each)

top_by_ratings(5, metric='median'): {}
258 μs ± 6.78 μs per loop (mean ± std. dev. of 3 runs, 100 loops each)

top_controversial(5): {}
259 μs ± 4.23 μs per loop (mean ± std. dev. of 3 runs, 100 loops each)


👥 Ratings.Users методы:

dist_by_num_of_rating: {}
954 ns ± 34.3 ns per loop (mean ± std. dev. of 3 runs, 100 loops each)

dist_by_rating_values(metric='average'): {}
868 ns ± 34.4 ns per loop (mean ± std. dev. of 3 runs, 100 loops each)

dist_by_rating_values(metric='median'): {}
844 ns ± 22.2 ns per loop (mean ± std. dev. of

## 🏷️ Класс `Tags`

> **Назначение:** Анализ данных из файла `tags.csv`  
> **Источник:** `./ml-latest-small/tags.csv`  
> **Объём выборки:** первые 1000 строк (плюс заголовок = 1001 строка)  
> **Время инициализации:** `253 μs ± 30.8 μs`


### 🏷️ Исследование концептуального слоя: Класс `Tags`

**Цель этапа** — провести комплексный анализ неструктурированных текстовых меток (тегов), оставленных зрителями к фильмам, и трансформировать субъективные пользовательские смыслы в четкие аналитические паттерны.

Мы начинаем погружение с текстового и понятийного пласта данных. Анализ класса `Tags` позволит решить следующие задачи:
1. Выявить скрытые пользовательские тренды и маркеры восприятия кино.
2. Структурировать хаотичные текстовые описания для их дальнейшей математической обработки.
3. Понять, какие смысловые акценты аудитория расставляет чаще всего.

---
> 💡 *Переходим от сырых, неструктурированных данных к поиску закономерностей и инсайтов.*


In [40]:
from movielens_analysis import Tags

In [41]:
folder = "./ml-latest-small/tags.csv"
tags = Tags(folder)
%timeit -n 1 -r 2 tags = Tags(folder)
print(f"Length: {len(tags.list_1000_strings)}")

2.11 ms ± 111 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Length: 1001


При инициализации класса Tags парсится файл tags.csv и берутся первые 1000 строк (без учета заголовков)

### 🏷️ Функционал класса `Tags`

В рамках разработанного класса реализован комплекс аналитических инструментов для трансформации хаотичных пользовательских меток в структурированные инсайты. Наш алгоритмический стек решает следующие задачи:

1.  **🧩 Многословные концепты (`most_words`):** Выделяет топ-$n$ тегов с наибольшим количеством отдельных слов. Помогает находить самые детализированные и развернутые описания фильмов.

2.  **📏 Максимальная длина тегов (`longest`):** Измеряет физическую длину меток в символах, возвращая топ-$n$ самых длинных строк, отсортированных по убыванию.

3.  **🔗 Пересечение текстовых гигантов (`most_words_and_longest`):** Находит общие элементы на стыке многословных и длинных тегов. Изолирует уникальные, максимально перегруженные смыслом формулировки.

4.  **🔥 Глас народа (`most_popular`):** Подсчитывает частоту упоминания каждого тега, наглядно демонстрируя главные тренды и доминирующие темы в обсуждениях.

5.  **🔍 Тематический поиск (`tags_with`):** Интеллектуальный фильтр для контекстного поиска уникальных меток по ключевому слову с алфавитной сортировкой.

---

### 🚀 Вопрос исследования: Выявление скрытых тем через инструмент `tags_with`

**💡 Ключевая гипотеза:** Можно ли с помощью длинных пользовательских меток оперативно определять глобальные смысловые тренды и собирать тематические кластеры?

*   **Проблема:** Ручной поиск фильмов конкретной тематики (например, о космических одиссеях или гангстерских разборках) через чтение тысяч аннотаций неэффективен.
*   **Решение:** Применение метода `tags_with`. Он мгновенно группирует уникальные теги по ключевому слову, формируя готовый тематический пласт.

* Поработаем с методами класса и сделаем вывод основных методов.
Будем выводить топ 5 для каждой функции топа.

*Ниже представлен код реализации и результат работы фильтра:*

In [42]:
print("--- ЗАМЕРЫ СКОРОСТИ МЕТОДОВ И РЕЗУЛЬТАТЫ ---")

print(f"\nmost_words: {tags.most_words(5)}")
%timeit -n 100 -r 3 tags.most_words(5)

print(f"\nlongest: {tags.longest(5)}")
%timeit -n 100 -r 3 tags.longest(5)

print(f"\nmost_words_and_longest: {tags.most_words_and_longest(5)}")
%timeit -n 100 -r 3 tags.most_words_and_longest(5)

print(f"\nmost_popular: {tags.most_popular(5)}")
%timeit -n 100 -r 3 tags.most_popular(5)

print(f"\ntags_with: {tags.tags_with('sci-fi')}")
%timeit -n 100 -r 3 tags.tags_with('sci-fi')

--- ЗАМЕРЫ СКОРОСТИ МЕТОДОВ И РЕЗУЛЬТАТЫ ---

most_words: {'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16, 'the catholic church is the most corrupt organization in history': 10, 'Oscar (Best Music - Original Score)': 6, 'Everything you want is here': 5, 'based on a true story': 5}
346 μs ± 157 μs per loop (mean ± std. dev. of 3 runs, 100 loops each)

longest: ['Something for everyone in this one... saw it without and plan on seeing it with kids!', 'the catholic church is the most corrupt organization in history', 'audience intelligence underestimated', 'Oscar (Best Music - Original Score)', 'assassin-in-training (scene)']
174 μs ± 2.39 μs per loop (mean ± std. dev. of 3 runs, 100 loops each)

most_words_and_longest: ['Oscar (Best Music - Original Score)', 'Something for everyone in this one... saw it without and plan on seeing it with kids!', 'the catholic church is the most corrupt organization in history']
441 μs ± 20.2 μs per loop (mean ±

### 📊 Аналитический вывод: Что открывают нам методы `most_words`, `longest`, `most_words_and_longest` и `most_popular`?

Совместное применение этих метрик позволяет разбить весь массив пользовательских тегов на два противоположных, но крайне важных для анализа пласта:

1.  **⚡️ Поверхностный срез («Глас народа» через `most_popular`):**  
    Показывает самые массовые, лаконичные и понятные всем тренды. Это «верхушка айсберга» — ключевые темы, которые лежат на поверхности и объединяют большинство зрителей.

2.  **🧠 Глубинный срез («Смысловые гиганты» через `most_words` + `longest` + `most_words_and_longest`):**  
    Позволяет заглянуть внутрь пользовательского восприятия. Эти методы изолируют не просто ключевые слова, а уникальные мини-рецензии и сложные концепты. Именно здесь скрываются самые точные, детализированные и нестандартные характеристики фильмов, которые никогда не попадут в топ популярных, но идеально отражают суть киноленты.

---
> **📌 Итог:** Гипотеза подтверждается. Комбинация популярности и текстовой сложности тегов позволяет анализировать аудиторию на двух уровнях: быстро определять глобальные коммерческие тренды и одновременно извлекать редкие, глубокие смысловые инсайты.


## 🔗 Класс `Links`

> **Назначение:** Обогащение данных о фильмах через веб-парсинг TMDB и агрегация финансовой/производственной статистики  
> **Источники:** `links.csv`, `movies.csv` (первые 1000 строк) + скрапинг `themoviedb.org`  
> **Особенность:** Работает с внешним веб-интерфейсом, кеширует извлечённые поля в `self.movie_data`

---

### 🔧 Инициализация

In [43]:
from movielens_analysis import Links
links = Links("./ml-latest-small/links.csv", "./ml-latest-small/movies.csv")

### 🌐 Архитектура связей: Класс `Links` как глобальный цифровой мост

На этом этапе мы переходим к интеграции данных. Класс `Links` выполняет критически важную роль **информационного моста** 🌉 между локальной экосистемой `MovieLens` и бескрайней глобальной интернет-энциклопедией мирового кинематографа.

### 🛠️ Что это дает нашей аналитике?

*   **🔗 Синхронизация миров:** Метод связывает внутренние идентификаторы датасета со строго верифицированными внешними базами данных.
*   **🌍 Выход в открытый интернет:** Через сквозные индексы `imdbId` и `tmdbId` мы получаем прямой доступ к колоссальным внешним ресурсам:
    *   `imdbId` для `IMDb` ➔  <http://www.imdb.com>
    *   `tmdbId` для `TMDb` ➔  <https://www.themoviedb.org>

---
> **🚀 Стратегическая цель:** Превратить изолированный учебный датасет в открытую аналитическую систему, готовую к обогащению любыми внешними данными из веб-пространства.


### В рамках этого класса успешно реализованы следующие аналитические задачи:

🌐 Парсинг данных (get_tmdb): Написан интеллектуальный алгоритм, который заходит на веб-страницы themoviedb, извлекает актуальную информацию (Режиссеров, Бюджеты, Кассовые сборы, Длительность) для каждого фильма и аккуратно сортирует базу по убыванию ID фильмов.

🎬 Топ-режиссёров мирового кино (top_directors): Создан инструмент, выявляющий самых продуктивных и влиятельных создателей кино по количеству снятых шедевров.

💰 Рейтинг киногигантов (most_expensive): Реализован поиск самых финансово затратных и амбициозных кинопроектов в истории.

📈 Анализ абсолютной окупаемости (most_profitable): Рассчитана чистая прибыль картин как разница между мировыми кассовыми сборами и бюджетом, что позволило выявить главных коммерческих триумфаторов индустрии.

⏱️ Измерение кинематографического времени (longest): Выделены самые эпичные и продолжительные киноленты, способные удерживать зрителя у экрана дольше всего.

💎 Индекс плотности инвестиций (top_cost_per_minute): Разработана уникальная метрика «цена минуты», которая показывает, сколько стоила каждая секунда экранного времени для продюсеров, и определяет самые дорогие в производстве фильмы с точностью до двух знаков после запятой.

### 🔗 Тестирование методов и замер скорости: класс `Links`

В данном блоке проводится проверка аналитических методов класса `Links` и замер времени их выполнения.

🔍 **Что тестируется:**
- Корректность работы методов агрегации: `top_directors`, `most_expensive`, `most_profitable`, `longest`, `top_cost_per_minute`
- Формат возвращаемых значений (`dict[str, int]` / `dict[str, float]`)
- Сортировка результатов по убыванию целевого метрика
- Стабильность работы с in-memory структурой `self.movie_data`

⚡ **Особенности замеров скорости:**
- Сетевой запрос `get_tmdb()` выполняется **один раз** перед замерами. Его время не учитывается в `%timeit`, так как зависит от скорости сети, лимитов TMDB и может сильно варьироваться.
- Все аналитические методы работают с уже загруженными в память данными, поэтому `%timeit` показывает стабильные значения в диапазоне **микросекунд**.
- Для избежания избыточного вывода в лог количество прогонов снижено до `-n 10 -r 3` (достаточно для оценки скорости на готовых данных).

📥 **Запуск:** выполните код в следующей ячейке. Вывод покажет топ-5 результатов для каждого метода и среднее время выполнения.

In [44]:
print("--- LINKS: ЗАМЕРЫ И РЕЗУЛЬТАТЫ ---\n")

# ⚠️ Сначала загрузим данные с TMDB (один раз, т.к. сеть)
%timeit -n 3 -r 1 links.get_tmdb(list(links.movie_data.keys())[:5], ["Director", "Budget", "Revenue", "Runtime"])

print(f"top_directors: {links.top_directors(5)}")
%timeit -n 10 -r 3 links.top_directors(5)

print(f"\nmost_expensive: {links.most_expensive(5)}")
%timeit -n 10 -r 3 links.most_expensive(5)

print(f"\nmost_profitable: {links.most_profitable(5)}")
%timeit -n 10 -r 3 links.most_profitable(5)

print(f"\nlongest: {links.longest(5)}")
%timeit -n 10 -r 3 links.longest(5)

print(f"\ntop_cost_per_minute: {links.top_cost_per_minute(5)}")
%timeit -n 10 -r 3 links.top_cost_per_minute(5)

--- LINKS: ЗАМЕРЫ И РЕЗУЛЬТАТЫ ---

9.5 s ± 0 ns per loop (mean ± std. dev. of 1 run, 3 loops each)
top_directors: {'John Lasseter': 1, 'Joe Johnston': 1, 'Howard Deutch': 1, 'Forest Whitaker': 1, 'Charles Shyer': 1}
42.3 μs ± 1.49 μs per loop (mean ± std. dev. of 3 runs, 10 loops each)

most_expensive: {'Jumanji (1995)': 65000000, 'Toy Story (1995)': 30000000, 'Grumpier Old Men (1995)': 25000000, 'Waiting to Exhale (1995)': 16000000, 'Father of the Bride Part II (1995)': 0}
26.9 μs ± 1.2 μs per loop (mean ± std. dev. of 3 runs, 10 loops each)

most_profitable: {'Toy Story (1995)': 371157969, 'Jumanji (1995)': 197821940, 'Father of the Bride Part II (1995)': 76594107, 'Waiting to Exhale (1995)': 65452156, 'Grumpier Old Men (1995)': 46518503}
99.5 μs ± 60 μs per loop (mean ± std. dev. of 3 runs, 10 loops each)

longest: {'Waiting to Exhale (1995)': 127, 'Father of the Bride Part II (1995)': 106, 'Jumanji (1995)': 104, 'Grumpier Old Men (1995)': 101, 'Toy Story (1995)': 81}
36.1 μs ± 13.

### 🔗 Тестирование динамического парсинга: `Links.get_tmdb()`

### 🎯 Цель теста
Проверить, что метод `get_tmdb()` корректно обрабатывает **произвольный набор полей**, передаваемых в параметре `list_of_fields`, и возвращает данные только по запрошенным ключам.

### 💻 Код теста

In [45]:
m_ids = list(links.movie_data.keys())[:5]

# Запрашиваем кастомный набор полей — смесь стандартных и расширенных
fields_to_parse = ['Director', 'Status', 'Original Language', 'Budget', 'Screenplay']

result = links.get_tmdb(m_ids, fields_to_parse)

print(f"✅ Запрошено полей: {fields_to_parse}")
print(f"✅ Обработано фильмов: {len(result)}")
print(f"\n📄 Результаты:")
for row in result:
    print(row)


✅ Запрошено полей: ['Director', 'Status', 'Original Language', 'Budget', 'Screenplay']
✅ Обработано фильмов: 5

📄 Результаты:
[5, 'Charles Shyer', 'Released', 'English', 0, 'Nancy Meyers']
[4, 'Forest Whitaker', 'Released', 'English', 16000000, 'Ronald Bass']
[3, 'Howard Deutch', 'Released', 'English', 25000000, 'N/A']
[2, 'Joe Johnston', 'Released', 'English', 65000000, 'Jim Strain, Jonathan Hensleigh, Greg Taylor']
[1, 'John Lasseter', 'Released', 'English', 30000000, 'Joel Cohen, Alec Sokolow, Andrew Stanton, Joss Whedon']


## 🏁 Заключение: Синергия кода, данных и качества

Наше аналитическое путешествие по вселенной **MovieLens** успешно завершено. В ходе работы над проектом мы спроектировали не просто набор разрозненных скриптов, а полноценную, гибкую и масштабируемую **аналитическую экосистему**. 

### ⚙️ Ключевые итоги нашего исследования:

*   **🧩 Модульная гармония:** Разделение предметной области на четыре изолированных класса (`Movies`, `Ratings`, `Tags`, `Links`) позволило разложить хаотичный массив данных на понятные составляющие. Мы связали воедино историческую хронологию, миллионы пользовательских эмоций, текстовые ассоциации и экономические показатели киноиндустрии.
*   **🧪 Культура промышленной разработки:** Интеграция класса `Test` на базе фреймворка **PyTest** доказала, что аналитика может и должна быть надежной. Покрытие кода модульными тестами гарантирует, что наши алгоритмы устойчивы к аномалиям, пропускам в данных и готовы к работе на любых новых объемах информации.
*   **⏱️ Эффективность и оптимизация:** Использование магических команд `%timeit` на протяжении всего отчета позволило нам контролировать скорость выполнения кода в реальном времени. Это гарантирует, что разработанные методы работают оптимально и готовы к масштабированию под Big Data.
*   **🎬 Инсайты на стыке дисциплин:** Нам удалось превратить сухие файлы логов в увлекательную историю кинематографа. Мы увидели, как тесно связаны между собой жанровые тренды, финансовые затраты продюсеров и живой отклик реальных зрителей.

Этот проект наглядно демонстрирует, что за каждым успешным рекомендательным алгоритмом современных стриминговых сервисов стоит строгая, математически выверенная архитектура и чистый, протестированный код. Разработанный инструментарий оставляет после себя надежный фундамент для построения будущих прогнозных моделей и рекомендательных нейросетей!
